### PR TITLE
Update dotty universe to have a context

### PR DESCRIPTION
**Pull request by liufengyun**
This PR also remove `GenSym` and introduce an abstract method `fresh` in Universe.

Only the last commit is nonw.

### CODE REVIEW

**Comment by olafurpg** [comment]
Can you please rebase on master? I merged #24

File: core/src/main/scala/scala/macros/core/Universe.scala [hunk 1]
```diff
@@ -26,12 +24,15 @@ trait Universe extends Gensym {
   type PatVar // NOTE(olafur) subject for removal
   type Defn
 
+  def fresh(prefix: String): String
```
Comment (olafurpg) [inline core/src/main/scala/scala/macros/core/Universe.scala:27]: Would it make sense to replace this with `TermNameFresh: TermName`?

**Review by olafurpg — APPROVED** [review]
LGTM 👍

File: core/src/main/scala/scala/macros/core/Universe.scala [hunk 2]
```diff
@@ -81,11 +82,11 @@ trait Universe extends Gensym {
...
-  def symName(sym: Symbol)(implicit m: Mirror): Name
-  def denotInfo(denot: Denotation)(implicit m: Mirror): Type
-  def denotName(denot: Denotation)(implicit m: Mirror): Name
-  def denotSym(denot: Denotation)(implicit m: Mirror): Symbol
-  def caseFields(tpe: Type)(implicit m: Mirror): List[Denotation]
+  def symName(sym: Symbol): Name
+  def denotInfo(denot: Denotation): Type
+  def denotName(denot: Denotation): Name
+  def denotSym(denot: Denotation): Symbol
+  def caseFields(tpe: Type): List[Denotation]
```
Comment (liufengyun) [inline core/src/main/scala/scala/macros/core/Universe.scala:89]: For another issue: `caseFields` is a little overly specific, we can separate it to `constructorFields` and `symbol.isCaseClass`.

File: core/src/main/scala/scala/macros/core/Universe.scala (hunk 1 above)
Comment (liufengyun) [reply core/src/main/scala/scala/macros/core/Universe.scala:27]: Seems it complicates the service protocol a little bit.

File: core/src/main/scala/scala/macros/core/Universe.scala (hunk 2 above)
Comment (olafurpg) [reply core/src/main/scala/scala/macros/core/Universe.scala:89]: I opened https://github.com/scalacenter/macros/issues/27

**Pull request merged**